design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,141 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to immediately propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out DaemonSets.
+
+
+## Goals
+
+-   Provide a way to propagate changes from a running VM object to its corresponding VMI immediately without interrupting the workload.
+    
+-   Maintain backwards compatibility
+    
+
+## Non Goals
+
+-   Successfully apply all the changes to the Virtual Machine Instance Object. Not all fields can be propagated to the guest immediately. Any change affecting the guest API will remain staged and will require a restart to get applied.
+    
+-   Make LiveUpdate the default. We defer this decision.
+    
+
+## Definition of Users
+
+A user - a VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a user I want changes that I applied to the VM object to propagate to the VMI immediately if possible.
+
+As a user, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+A new field called vmRolloutStrategy will be introduced in the KubeVirt CR. This field will control the behavior of the VM objects. Two possible values will be available:
```

Comment:
Is the "stage" strategy there because we think it is a valid and useful use case or we keep it just because the live update is not graduated and we are afraid there are some automation scripts depend on the old behavior?
If the answer is the second option, then I think it perfectly fits a feature gate.